### DISCUSSION TITLE
Trouble implementing Dirichilet Boundary Conditions in a gmsh with Firedrake

### DISCUSSION THREAD

**Discussion by Arijit-Majumdar** [firedrakeproject/firedrake#3406] [Firedrake support] [answered]
Hi,
I am trying to use Firedrake in colab with gmsh for a 2D Poisson equation. The geometry is a square domain with 5 holes in it. Neumann BC is applied at the outer edges of the domain and Dirichilet BC has to be applied at the inner holes. Here is how the mesh looks like
![MeshImage](https://github.com/firedrakeproject/firedrake/assets/31143600/e0fefb32-5f47-47d6-803b-d015fdbe08c6)

But while trying to include the Dirichlet BC and solving the system I keep getting the following error
`LookupError: BC construction got invalid markers {5}. Valid markers are 'set()'`

I am providing a link to my colab notebook
https://colab.research.google.com/drive/1SOaM9EuobCSF15r2t3nEyYb6tPQgYW_u?usp=sharing

Any help will be greatly appreciated.

Thanks
Arijit

**Comment by dham** [comment, marked as answer]
Looks like your Gmsh mesh does not have boundary markers. You need Gmsh "Physical" entities in order to attach the boundary markers that Firedake will see. Please see this demo: https://www.firedrakeproject.org/demos/immersed_fem.py.html

Note also that we can't see your notebook, because it's not publicly shared. It's far better to create a minimal failing example and put it directly into your GitHub post inside code fences (put ```python on the line before your code and ``` on the line after it).

**Reply by Arijit-Majumdar** [reply]
Thanks for your response. I managed to include Physical entities in my mesh. However I am getting the following error
`ValueError: Provided Jacobian is not a bilinear form`

Here is how my bilinear and linear functional look like

`a = 0.0002*dot(grad(v),grad(u))*dx`
`L = Constant(fe)*v*dx + Constant(h1)*v('+')*dS(1) + Constant(h2)*v('+')*dS(2) + Constant(h3)*v('+')*dS(3) + Constant(h4)*v('+')*dS(4)`

Here 1, 2, 3, 4 are the physical group labels for the four edges of the square domain with Neumann BC.
As for the Dirichilet BC, here is what I am doing

`DirBC1 = DirichletBC(V, g1, [5])`
`DirBC2 = DirichletBC(V, g2, [6,7,8,9])`

Finally I am solving the system as follows
`solve(a == L, u, bcs=[DirBC1,DirBC2], solver_parameters={'ksp_type': 'cg'})`

**Reply by connorjward** [reply]
Please share the full (minimal) code snippet. Your example doesn't show us the types of `u` and `v` (and more) so we can't reproduce your error.

**Reply by Arijit-Majumdar** [reply]
Hi, here is a link to my notebook
https://colab.research.google.com/drive/1SOaM9EuobCSF15r2t3nEyYb6tPQgYW_u?usp=sharing

The u and v are as follows
`V = FunctionSpace(mesh, "CG", 1)`
`u = TrialFunction(V)`
`v = TestFunction(V)`

These are the constants I am using 
`g1=2500`
`g2=100`
`h1=0.9`
`h2=0`
`h3=-0.1`
`h4=0`
`h5=0`
`h6=0`
`fe = 0.2`

Please let me know if you need anything else from my side.
Thanks for the help.

**Reply by connorjward** [reply]
I was able to reproduce your error by running the cells out of turn. You redefine `u` from `TrialFunction(V)` to `u = Function(V)` and using the latter in your definition of `a` will break things. I am able to run the solve by making sure to execute the cells in order.
